nexus/src/app/background/instance_watcher.rs

```diff
@@ -0,0 +1,593 @@
+                        datum: Cumulative::default(),
+                    },
+                    CheckOutcome::Success(state) => Check {
+                        state: state.to_string(),
+                        reason: "success".to_string(),
+                        datum: Cumulative::default(),
+                    },
+                })
+                .datum += 1;
+
+            self.touched = true;
+        }
+
+        pub(super) fn check_error(&mut self, reason: Incomplete) {
+            self.check_errors
+                .entry(reason)
+                .or_insert_with(|| IncompleteCheck {
+                    reason: reason.to_string(),
+                    datum: Cumulative::default(),
+                })
+                .datum += 1;
+            self.touched = true;
+        }
+
+        fn len(&self) -> usize {
+            self.checks.len() + self.check_errors.len()
+        }
+
+        fn sample_into(
+            &self,
+            target: &VirtualMachine,
+            dest: &mut Vec<Sample>,
+        ) -> Result<(), MetricsError> {
+            for metric in self.checks.values() {
+                dest.push(Sample::new(target, metric)?);
+            }
+            for metric in self.check_errors.values() {
+                dest.push(Sample::new(target, metric)?);
+            }
+            Ok(())
+        }
+    }
+
+    /// The number of successful checks for a single instance, VMM, and sled agent.
+    #[derive(Clone, Debug, Metric)]
+    struct Check {
+        /// The string representation of the instance's state as understood by
+        /// the VMM. If the check failed, this will generally be "failed".
+        state: String,
+        /// `Why the instance was marked as being in this state.
+        ///
+        /// If an instance was marked as "failed" due to a check failure, this
+        /// will be a string representation of the failure reason. Otherwise, if
+        /// the check was successful, this will be "success". Note that this may
+        /// be "success" even if the instance's state is "failed", which
+        /// indicates that we successfully queried the instance's state from the
+        /// sled-agent, and the *sled-agent* reported that the instance has
+        /// failed --- which is distinct from the instance watcher marking an
+        /// instance as failed due to a failed check.
+        reason: String,
+        /// this will be a string representation of the failure reason.
+        datum: Cumulative<u64>,
+    }
+
+    /// The number of unsuccessful checks for an instance and sled agent pair.
+    #[derive(Clone, Debug, Metric)]
+    struct IncompleteCheck {
+        /// The reason why the check was unsuccessful.
+        ///
+        /// This is generated from the [`Incomplete`] enum's `Display` implementation.
+        reason: String,
+        /// The number of failed checks for this instance and sled agent.
+        datum: Cumulative<u64>,
+    }
```

Comment:
Here is where I could use a little advice on the Oximeter idioms: it's occurred to me that we could potentially roll these into a single metric, now that the `Check` metric has separate "state" and "reason" fields. I wanted to make sure that the metric structure could represent both "check has marked the instance as failed due to this error" and "check has not made a determination about the instance due to an internal check error", so I've started by having two separate metrics `Check` and `IncompleteCheck`. However, it occurs to me that incomplete checks _could_, alternatively, be represented by a `state` string like `state: "unknown"` or `state: "incomplete"`, with the `reason` field indicating why we weren't able to do a check successfully. 

@bnaecker (or someone else who's good at Oximeter), what do you think? Is it better to emit more separate timeseries, to represent these, or combine them into one that can represent both cases?

Reply:
This is a good question, and though I've been thinking for a while, I'm very sorry to say I don't have a good answer for you :/

In general, I'd probably make this decision based on how one wants to query the data. If one wants to easily break out or group by a field, they should probably be part of the same timeseries. If that's rare, then I'd probably put these mostly-unrelated metrics (instance-state vs instance-state-fetch-failures) into two different timeseries.

My only hesitation with grouping them is that it sort of complicates the meaning of the data. IIUC, then the `reason` field is non-empty iff the state is `"unknown"`, i.e., we failed the check. So even the successful samples carry this now-empty field. To be clear, I'm not worried about resource usage here, just confusion about what the field means, and when it's valid. Maybe changing that field name to `failure_reason` is enough to mitigate that confusion, though. And this is a pretty non-specific worry -- I think it should mechanically work fine to do that, even if one wants to separately query for things like known states or failed checks. In either case, we can answer the same kinds of questions:

- what fraction of time was the instance in some state (say, stopped or running), over some period of time?
- how often did the instance change states?
- was the instance rebooted in this period?
- why did Nexus fail to determine the instance's state?

So I'd just say, choose what feels most natural to you at this point! I wish I had more concrete advice for you, but I guess we'll have to develop that together as we use it more :)

Reply:
@bnaecker 
> IIUC, then the reason field is non-empty iff the state is `"unknown"`, i.e., we failed the check. So even the successful samples carry this now-empty field. To be clear, I'm not worried about resource usage here, just confusion about what the field means, and when it's valid. Maybe changing that field name to failure_reason is enough to mitigate that confusion, though.

So, it's a little bit more complex than that, actually. There are (currently) three possible outcomes for an instance check:
- The check was "successful", indicating that we got a state back from the sled-agent (which may be "failed", if the sled-agent has decided that the instance has failed)
- The check was a "failure", indicating that we got an _error_ from the sled-agent that affirmatively indicates that the instance has failed. Presently, this occurs when we get back an HTTP error indicating that the sled agent doesn't know that it it has the instance we expected it to know about, which means that the sled-agent has crashed and the instance is now lost, or if the sled-agent was totally unreachable.
- The check was "incomplete", indicating that it failed in a way that made it impossible for us to determine what was up with the instance. This includes HTTP client errors and other things that indicate the *nexus* process messed up.

In the first two cases, successful and failed checks, we will emit both a "state" field, representing the determination about the instance's state, and a "reason" field, indicating _why_ we made that determination, which will be "success" if the check succeeded with an affirmative response from the sled-agent[^1], or a representation of the error that caused the check to fail. In the incomplete case, we instead increment the `virtual_machine:incomplete_check` metric, indicating that we couldn't figure out what was going on. 

What I'm proposing here is representing the "incomplete check" case using the same `virtual_machine:check` metric, instead, by emitting "unknown" as the instance state.

Reply:
Ok, thanks for clarifying. I still don't have a great answer, and thing both approaches have their merits. Keeping them separate seems reasonable because the second metric really tracks a failure to communicate with the sled agent for some reason, but complicates checking the instance state itself in that case (unknown). Merging them makes tracking the state simpler, but complicates the meaning of the metric itself (it includes information about communication with the sled agent).

I think I have a slight preference for merging them, as you're suggesting. However, I think if we do keep them separate, I would emit a metric for the instance state as "unknown" in addition to the metric describing why the check failed. That is, for every sample period we should have a new value for the state of the instance itself, even if it's "unknown".

Reply:
I think I'm going to go ahead and merge them, then, based on the rationale that if we emit a `state: "unknown"` for every incomplete check, then the `incomplete_check` metric isn't really capturing any information we don't already have.

Reply:
Actually, upon further consideration, I think having a separate `incomplete_check` metric _is_ the right thing, because it captures additional information about another possible kind of incomplete check: one where we _do_ determine the instance's state, but fail to update the database for that instance. In that case, we would want to record the instance state we determined, but _also_ record that we didn't successfully update the instance in CRDB. I think that emitting a `state: "unknown"` for any check that _didn't_ successfully determine the state is a good idea though.